conan/tools/b2/b2deps.py

```diff
@@ -0,0 +1,284 @@
+
+    def generate(self):
+        """
+        This method will save the generated files to the conanfile.source_folder
+        """
+        source_folder = self._conanfile.source_folder
+        self._conanfile.output.highlight(f"Writing B2Deps to {source_folder}")
+        with chdir(source_folder):
+            check_duplicated_generator(self, self._conanfile)
+            generator_files = self.content
+            for generator_file, content in generator_files.items():
+                self._conanfile.output.info(f"Saved B2Deps file {generator_file}")
+                save(generator_file, content)
+
+    @property
+    def content(self):
+        """
+        Generates two content files: conanbuildinfo.jam and
+        conanbuildinfo-ID.jam. The former defines common package definition
+        function and include the latter conanbuildinfo-ID.jam files. The
+        conanbuildinfo-ID.jam files define sub-projects and targets for each
+        settings variation. The generated files have stable names and content.
+        Hence they can be added to source control.
+        """
+        self._content = {}
+        self._content_conanbuildinfo_jam()
+        self._content_conanbuildinfo_variation_jam()
+        for ck in self._content.keys():
+            self._content[ck] = self._conanbuildinfo_header_text+"\n"+self._content[ck]
+        return self._content
+
+    def _content_conanbuildinfo_jam(self):
+        # Generate the common conanbuildinfo.jam which does four things:
+        #
+        # Defines common utility functions to make the rest of the code short
+        # and includes the conanbuildinfo-*.jam sub-files.
+        cbi = [self._conanbuildinfo_common_text]
+        # The combined text.
+        self._content['conanbuildinfo.jam'] = "\n".join(cbi)
+
+    def _content_conanbuildinfo_variation_jam(self):
+        # Generate the current build variation conanbuildinfo-/variation/.jam.
+        for require, dependency in self._conanfile.dependencies.items():
+            # Only generate defs for direct dependencies.
+            if not require.direct:
+                continue
+            # The base name of the dependency.
+            dep_name = dependency.ref.name
+            # B2 equivalent of the dependency name. We keep all names lower case.
+            dep_name_b2 = dep_name.lower()
+            # The dependency cpp_info. We need to consider that there's a
+            # "_depname" component. Such components are a kludge to appease
+            # cmake generators and will eventually go away. This special
+            # component holds the real root definitions of the dependency.
+            dep_cpp_info = dependency.cpp_info
+            if '_'+dep_name in dep_cpp_info.components:
+                dep_cpp_info = dep_cpp_info.components['_'+dep_name]
+            # The settings and options the dependency requires, i.e. finds relevant.
+            dep_settings = dependency.settings
```

Comment:
All the other generators don't use the settings of the dependency, but instead the settings of the consumer. When something like ``conan install -s &:build_type=Debug -s build_type=Release`` the generated files have a "_debug" in their name, it means: for the "debug" build of the consumer, use this (and it can be a release artifact, it doesn't matter).

Reply:
This still doesn't hold. Take into account:
```python
name = "mypkg"
def requires(self):
      if self.settings.build_type == "Release":
               self.requires("myalloc/2.0")
      else:
               self.requires("myalloc/1.0")
```
(I have seen this in production code, more than once)

The settings are the settings of the consumer, and should match just 1 config. It is better not to try to implement a ``multi-config`` approach within a single ``conan install``, or it will have other unexpected consequences.

Reply:
Why wouldn't this work, though? You would have 2 completely different targets in `/myalloc` project with these two installs. With

```
exe myexe : main.cpp /myalloc//myalloc ;
```
and  building with `b2 variant=release` the build sustem would choose one, and when building with `b2 variant=debug`, it would choose another.

Reply:
The only way that the dependency graph will point to ``myalloc/1.0`` is executing a ``conan install -s build_type=Debug``. It is completely impossible get the 2 different dependency graphs in one single ``conan install`` command, like ``myalloc/1.0`` will not even be in the computer if not executed. So:

> For example your ``conan install -s &:build_type=Debug -s build_type=Release would also produce the same results as
> conan install -s build_type=Debug
conan install -s build_type=Release

Cannot hold, it is not physically possible.

Reply:
Oh, sure, that's true. But I don't think that it means this multi-config setup doesn't work. You just have to run `conan install` several times.